Core/Src/main.c

```diff
@@ -115,36 +145,45 @@ int main(void)
   while (1)
   {
     /* USER CODE END WHILE */
-	  sendData(TxData);
-	  HAL_Delay(10);
-	  if (flag == 1 && (currentMillis - previousMillis > 20))
-	  {
-	    TxData[0] = 1;
-	    sendData(TxData);
-	    previousMillis = currentMillis;
-	    TxData[0] = 0;
-	  }
-	  if (flag == 2 && (currentMillis - previousMillis > 20))
-	  {
-	    TxData[0] = 2;
-	    sendData(TxData);
-	    previousMillis = currentMillis;
-	    TxData[0] = 0;
-	  }
-	 if (flag == 3 && (currentMillis - previousMillis > 20))
-	 {
-	    TxData[0] = 3;
-	    sendData(TxData);
-	    previousMillis = currentMillis;
-	    TxData[0] = 0;
-	  }
-	  if (flag == 4 && (currentMillis - previousMillis > 20))
-	  {
-	    TxData[0] = 4;
-	    sendData(TxData);
-	    previousMillis = currentMillis;
-	    TxData[0] = 0;
-	  }
+
+	 			  sendData(TxData);
+	 			 		 // if (led == RED_L)
+	 			  	  	  if(bit & (1 << 0))
```

Comment:
What do these bit-banging operations do?

Reply:
Joel wanted me to use a 32-bit (uint32) to hold each button state as single bit (up to 32 buttons), so the GPIO interrupt/timer callback uses a bitmask to set the corresponding bit. Then in the while loop I used a bitmask to read the corresponding bit for each relay/button. I then cleared the bit using a bitmask again once it was read. 

For example if button 1 (0th bit) was pressed, the interrupt would set the bit to a 1:
       0x0000 
|      0x0001
-----------------
= 0x0001
Then in the while loop, each if statement would check for a specific button:
      0x0001
&   0x0001
---------------
= 0x0001                 (which would make that statement true)
Then once detected, the if statement would be entered and send the appropriate data as well as clear the specific bit:
        0x0001
&  ~0x0001            
----------------
= 0x0000

Reply:
Sure, that's an interesting point. I will work on it!